bench: Make WalletLoading benchmark run faster

achow101 opened the pull request:
`minEpochIterations` is probably unnecessary to set, so removing it makes the runtime much faster.

achow101 committed:
bench: Remove minEpochIterations from wallet loading benchmark

This is probably unnecessary and just makes it slower.

MarcoFalke commented:
review ACK 9e404a98312d73c969adf4f8e87aad1ac4b3029d

w0xlt commented:
On my machine:

Before PR #24913
```
Executed in   57.34 secs    fish           external
   usr time   57.60 secs  843.00 micros   57.60 secs
   sys time    1.43 secs    0.00 micros    1.43 secs
```
After PR #24913
```
|    1,525,136,424.90 |                0.66 |    0.2% |    189.67 | `WalletLoadingDescriptors`
|      873,902,969.45 |                1.14 |    0.3% |    106.20 | `WalletLoadingLegacy`

________________________________________________________
Executed in  564.01 secs    fish           external
   usr time  540.03 secs  427.00 micros  540.02 secs
   sys time   13.35 secs   95.00 micros   13.35 secs
```
This PR
```
|    1,740,638,493.00 |                0.57 |    0.6% |     26.13 | `WalletLoadingDescriptors`
|    1,016,725,382.00 |                0.98 |    0.9% |     12.72 | `WalletLoadingLegacy`

________________________________________________________
Executed in  348.35 secs    fish           external
   usr time  328.66 secs    1.05 millis  328.66 secs
   sys time   12.46 secs    0.60 millis   12.46 secs
```

There's a good improvement  (348.35 secs) over PR #24913 (540.03 secs), but it still takes a lot longer than before it (57.34 secs).

fanquake commented:
> @fanquake Is this good enough or do you want faster?

I think to be good enough, it needs to run in a similar time to all of the other benchmarks, i.e max ~1-2 seconds, otherwise anyone running `make check` on macOS is going to be opening issues asking why running `make check` (has suddenly started) taking 10+ minutes. I tested this branch, but stopped after `WalletLoadingDescriptors` took 12 minutes:
```bash
|           80,237.38 |           12,463.02 |    0.1% |      0.01 | `VerifyScriptBench`
|            8,190.91 |          122,086.52 |    1.2% |      0.01 | `WalletBalanceClean`
|          179,780.20 |            5,562.35 |    1.5% |      0.01 | `WalletBalanceDirty`
|            7,929.42 |          126,112.61 |    0.1% |      0.01 | `WalletBalanceMine`
|               42.47 |       23,546,683.62 |    0.0% |      0.01 | `WalletBalanceWatch`
|      902,283,977.00 |                1.11 |    1.4% |    105.48 | `WalletLoadingDescriptors`
^C
src/bench/bench_bitcoin  59.66s user 24.31s system 11% cpu 12:01.03 total
```

> -unsafesqlitesync does not appear to have any effect. The next thing would be to reduce the number of txs created, but I'm concerned that doing so will make it harder to observe actual performance issues that this benchmark is intended to measure.

Given that, it seems like we've got another macOS only-performance issue, that we should probably figure out the root cause of?

achow101 committed:
walletdb: Create a mock database of specific type

We may want to make a mock database of either SQLite or BDB, not just
whatever the compiled default is.

achow101 committed:
bench: Enable loading benchmarks depending on what's compiled

Add descriptor wallet benchmark only if sqlite is compiled. Add legacy
wallet benchmark only if bdb is compiled.

achow101 committed:
bench: use unsafesqlitesync in wallet loading benchmark

achow101 committed:
bench: Use mock wallet database for wallet loading benchmark

Using in-memory only databases speeds up the benchmark, at the cost of
real world accuracy.

laanwj commented:
> I think to be good enough, it needs to run in a similar time to all of the other benchmarks, i.e max ~1-2 seconds, 

Agree. It's unacceptable to have a benchmark take in the order of minutes. Remember that benchmarks are there to measure small timespans (such as inner loop functions), not very long processes.

I've also seen a CI timeout due to this. Remember that we do `bench/bench_bitcoin &> /dev/null` in one of the CI runs.

achow101 committed:
bench: Add transactions directly instead of mining blocks

achow101 committed:
sqlite: Use in-memory db instead of temp for mockdb

The mock db can be in-memory rather than just at temp file.

achow101 committed:
bench: reduce the number of txs in wallet for wallet loading bench

achow101 committed:
bench: reduce number of epochs for wallet loading benchmark

achow101 commented:
I've added the following changes:
* The number of benchmark epochs is reduced from the default 11 to 5.
* Use `unsafesqlitesync`
* Use mock databases
  * Update sqlite's mock database implementation to use in-memory dbs rather than just temp dbs
* Instead of mining blocks to wallet addresses, we make transactions and add them directly with `AddToWallet`.
* Reduced the number of txs added to the wallet from 5000 to 1000.

This still doesn't make it as fast as the other benchmarks though.

***

Before #24913:

```
________________________________________________________
Executed in   11.37 secs    fish           external
   usr time   11.36 secs    0.00 micros   11.36 secs
   sys time    7.62 secs  435.00 micros    7.62 secs
```

After #24913:

```
________________________________________________________
Executed in  150.77 secs    fish           external
   usr time  148.99 secs  334.00 micros  148.99 secs
   sys time    9.42 secs   88.00 micros    9.42 secs
```

This PR:

```
________________________________________________________
Executed in   20.25 secs    fish           external
   usr time   20.21 secs  319.00 micros   20.21 secs
   sys time    7.92 secs  158.00 micros    7.92 secs
```

sipa commented:
This PR makes `./src/bench/bench_bitcoin -filter=WalletLoadingDescriptors` go from taking 246s to 2.3s for me, making "make check" a lot more usable.

(Ubuntu 22.04 on a Ryzen 5950X system; my ZFS filesystem setup may make flushing/syncing relatively slower)

ajtowns commented:
Haven't reviewed the code at all, but e673d8b475995075b696208386c9e45ae7ca3e20 reduces the runtime of the WalletLoadingDescriptors test from 4m25s to 0.1s for me

theStack reviewed (commented):
Concept ACK

Didn't look much at the code yet, but ran the benchmarks for both legacy and descriptor wallet on my (rather slow) system after each commit and master via

```
$ time ./src/bench/bench_bitcoin -filter=WalletLoadingLegacy
$ time ./src/bench/bench_bitcoin -filter=WalletLoadingDescriptors
```

with the following results:

|     |  Legacy    |  Descriptors
|-----|------------|-----------------
|master | 3m18.39s   | 5m40.95s
|**Remove minEpochIterations** (9e404a98312d73c969adf4f8e87aad1ac4b3029d) | 1m54.15s   | 2m51.50s
|use unsafesqlitesync (817c051364208d3f9e7e2af5700bd2bee5c9f303) | 1m53.68s   | 2m46.76s
|reduce number of epochs (d94244c4bf37365272a16eb2ce6517605b4c8a47) | 1m49.12s   | 2m38.29s
|add txs directly instead of mining blocks (f85b54ed27bd6eddb1e7035db02d542575b3ab24) | 1m27.61s   | 1m51.24s
|**reduce number of txs in wallet** (7c0d34476df446e3825198b27c6f62bba4c0b974) | 0m19.83s   | 0m27.06s
|create mock db (a1080802f8d7c3d1251ec6f2be33031f568deafa) | 0m19.56s   | 0m26.18s
|use in-memory db for mock db (49910f255f77e14fccf189353d188efac00d1445) | 0m19.36s   | 0m26.03s
|**use mock db** (4af3547ebac672a2d516e8696fd3580a766c27eb) | 0m09.16s   | 0m07.25s
load bench depending on what's compiled (e673d8b475995075b696208386c9e45ae7ca3e20)   | 0m09.51s   | 0m07.19s

Marked the commits in bold where the most gains (at least 2x) were observed. Looking closer at the descriptor wallet benchmark, if I (manually) revert commits 817c051364208d3f9e7e2af5700bd2bee5c9f303 (using unsafesqlitesync) and 49910f255f77e14fccf189353d188efac00d1445 (use in-memory mode for mock db) on the top commit via the patch

```diff
diff --git a/src/bench/wallet_loading.cpp b/src/bench/wallet_loading.cpp
index f61138378..994811c79 100644
--- a/src/bench/wallet_loading.cpp
+++ b/src/bench/wallet_loading.cpp
@@ -85,7 +85,6 @@ static std::unique_ptr<WalletDatabase> DuplicateMockDatabase(WalletDatabase& dat
 static void WalletLoading(benchmark::Bench& bench, bool legacy_wallet)
 {
     const auto test_setup = MakeNoLogFileContext<TestingSetup>();
-    test_setup->m_args.ForceSetArg("-unsafesqlitesync", "1");

     WalletContext context;
     context.args = &test_setup->m_args;
diff --git a/src/wallet/walletdb.cpp b/src/wallet/walletdb.cpp
index 72d483e61..26908b377 100644
--- a/src/wallet/walletdb.cpp
+++ b/src/wallet/walletdb.cpp
@@ -1203,7 +1203,7 @@ std::unique_ptr<WalletDatabase> CreateMockWalletDatabase(DatabaseOptions& option

     if (format == DatabaseFormat::SQLITE) {
 #ifdef USE_SQLITE
-        return std::make_unique<SQLiteDatabase>(":memory:", "", options, true);
+        return std::make_unique<SQLiteDatabase>("", "", options, true);
 #endif
         assert(false);
     }
```

and got a run-time of 0m07.41s. From that result, I'd conclude that commits 817c051364208d3f9e7e2af5700bd2bee5c9f303 and 49910f255f77e14fccf189353d188efac00d1445 barely have any effect on the run-time and could in theory be removed from this PR (no hard feelings though).

Rspigler commented:
tACK e673d8b475995075b696208386c9e45ae7ca3e20

This solves the `make check` issue with `bench_bitcoin` reported [here](https://github.com/bitcoin/bitcoin/issues/25185)

furszy commented on src/bench/wallet_loading.cpp:
```diff
@@ -72,21 +94,30 @@ static void WalletLoading(benchmark::Bench& bench, bool legacy_wallet)
+        wallet = BenchLoadWallet(std::move(database), context, options);
 
         // Cleanup
+        database = DuplicateMockDatabase(wallet->GetDatabase(), options);
```
In 4af3547e:

I'm not so sure why do we need to update the database here, guess that what you wanted to do is rollback the db to the first mocked database created at line 103 and not the recently loaded one? So in case of any modification, we ensure that the benchmark is always running from the same static db input?

achow101 replied:
Actually it's because deleting the wallet so that we can load it again will delete the database.

furszy replied:
k, that is because of the db destructor, good.

furszy reviewed (commented):
Code review ACK e673d8b4, nice PR.

With a note:
I think that we could keep generating a chain instead of adding the dummy txs directly, so the benchmark is more accurate to what happens in a real wallet (otherwise we are skipping some flows during the `LoadToWallet(tx)` process like the tx initial state update that requires a chain lookup). 

What do you think about generating the chain manually instead of using the whole `generatetoaddress` flow, so we skip the entire `BlockAssembler::CreateNewBlock`, follow up `TestBlockValidity`, finding the correct nonce etc.. that are surely pretty slow.
 
As the wallet doesn't care about chain validity, we directly could: create, concatenate and append blocks into the chain index manually, which should be pretty fast, so the wallet can use them during the load flow.

achow101 commented:
> if I (manually) revert commits [817c051](https://github.com/bitcoin/bitcoin/commit/817c051364208d3f9e7e2af5700bd2bee5c9f303) (using unsafesqlitesync)

I think `-unsafesqlitesync` has more of an effect on macs than linux?

glozow commented on src/wallet/walletdb.cpp:
```diff
@@ -1187,13 +1187,36 @@ std::unique_ptr<WalletDatabase> CreateDummyWalletDatabase()
+#ifdef USE_BDB
     return std::make_unique<BerkeleyDatabase>(std::make_shared<BerkeleyEnvironment>(), "", options);
 #endif
+    assert(false);
```
In a1080802f8d7c3d1251ec6f2be33031f568deafa "walletdb: Create a mock database of specific type ": I hit this line in testing with both `USE_SQLITE` and `USE_BDB` off? I may have just messed something up, but does this need e673d8b475995075b696208386c9e45ae7ca3e20 "Enable loading benchmarks depending on what's compiled " squashed in?

glozow reviewed (commented):
Concept ACK e673d8b475995075b696208386c9e45ae7ca3e20. For each commit, verified that there was a performance improvement without negating the purpose of the bench, and made some effort to verify that the code is correct.

Would be nice to get this merged. I haven't used `make check` without quitting at bench in months :(

Nit: could you update the OP to include the list of things you've changed to make it run faster? It currently only mentions `minEpochIterations`.

>> if I (manually) revert commits [817c051](https://github.com/bitcoin/bitcoin/commit/817c051364208d3f9e7e2af5700bd2bee5c9f303) (using unsafesqlitesync)

> I think -unsafesqlitesync has more of an effect on macs than linux?

Fwiw, I don't know exactly what `-unsafesqlitesync` does, but tried dropping the commit on my mac and ubuntu machines. It seems like neither had any change (specifically on my machines though, not sure about others).
